gobblin-data-management/src/main/java/org/apache/gobblin/data/management/version/finder/AbstractDatasetVersionFinder.java

```diff
@@ -86,6 +87,110 @@ public Collection<T> findDatasetVersions(Dataset dataset) throws IOException {
     return dataSetVersions;
   }
 
+  /**
+   * Find dataset version in the input {@link org.apache.gobblin.dataset}. Dataset versions are subdirectories of the
+   * input {@link org.apache.gobblin.dataset} representing a single manageable unit in the dataset.
+   *
+   * @param dataset {@link org.apache.gobblin.dataset} to directory containing all versions of a dataset
+   * @return - Returns an iterator for fetching each dataset version found.
+   * @throws IOException
+   */
+  @Override
+  public RemoteIterator<T> findDatasetVersion(Dataset dataset) throws IOException {
+    FileSystemDataset fsDataset = (FileSystemDataset) dataset;
+    Path versionGlobStatus = new Path(fsDataset.datasetRoot(), globVersionPattern());
+    return getDatasetVersionIterator(fsDataset.datasetRoot(), getRegexPattern(versionGlobStatus.toString()));
+  }
+
+  /**
+   * Returns an iterator to fetch the dataset versions for the datasets whose path {@link org.apache.hadoop.fs.Path}
+   * starts with the root and matches the globPattern passed
+   *
+   * @param root - Path of the root from which the Dataset Versions have to be returned
+   * @param pathPattern - Pattern to match the dataset version path
+   * @return - an iterator of matched data versions
+   * @throws IOException
+   */
+  public RemoteIterator<T> getDatasetVersionIterator(Path root, String pathPattern) throws IOException {
+    Stack<RemoteIterator<FileStatus>> iteratorStack = new Stack<>();
+    RemoteIterator<FileStatus> fsIterator = fs.listStatusIterator(root);
+    iteratorStack.push(fsIterator);
+    return new RemoteIterator<T>() {
+      FileStatus nextFileStatus = null;
+      boolean isNextFileStatusProcessed = false;
+
+      @Override
+      public boolean hasNext() throws IOException {
+        if (iteratorStack.isEmpty()) {
+          return false;
+        }
+        // No need to process if the next() has not been called
+        if (nextFileStatus != null && !isNextFileStatusProcessed) {
+          return true;
+        }
+        nextFileStatus = fetchNextFileStatus(iteratorStack, pathPattern);
+        isNextFileStatusProcessed = false;
```

Comment:
Is this iterator thread safe?
Just check can you use `Iterators.transform()` from guava here?

Reply:
I have moved the nextfileStatus and isNextFileStatusProcessed as a pair, to avoid partial updates of these 2 variables. Iterators are usually not thread-safe, the caller should avoid sharing iterator among threads. 
`Iterators.transform()` is mostly used when we have a collection and want to iterate over it. In our case, if we don't want to bring in a collection of dataset in-memory. We can still pass this iterator to the transform, is that what you are referring here ? Any advantage we get due to it ?